lib/trento/domain/sap_system/sap_system.ex

```diff
@@ -633,6 +703,36 @@ defmodule Trento.Domain.SapSystem do
     end
   end
 
+  defp maybe_emit_sap_system_deregistration_events(
+         %SapSystem{
+           sap_system_id: sap_system_id,
+           application: %Application{
+             instances: instances
+           }
+         },
+         %DeregisterApplicationInstance{
+           sap_system_id: sap_system_id,
+           deregistered_at: deregistered_at
+         }
+       ) do
+    if !Enum.any?(instances, fn %Instance{features: features} ->
+         String.contains?(features, "MESSAGESERVER")
+       end) ||
+         !Enum.any?(instances, fn %Instance{features: features} ->
+           String.contains?(features, "ABAP")
+         end) do
```

Comment:
Maybe a bit cleaner with the help of variables, `unless` and text-match operator:

```
    has_abap? = Enum.any?(instances, fn instance -> instance.features =~ "ABAP" end)

    has_messageserver? =
      Enum.any?(instances, fn instance -> instance.features =~ "MESSAGESERVER" end)

    unless has_abap? && has_messageserver? do
    ....
```

Reply:
Other solution could be:
```
# We map all the available features
instances
|> Enum.map(fn %{features: features} -> features end)
|> Enum.uniq()

# Check if message server and abap ara available in our list checking if the subset is there (https://stackoverflow.com/questions/63143363/elixir-check-array-contains-all-the-values-of-another-array)
unless Enum.empty?(["MESSAGESERVER", "ABAP"] -- features) do
   ...
end
```